### ISSUE TITLE
Rule proposal: Code files should (or should not) end with an empty line

### ISSUE THREAD

**Issue by NikolayIT** [DotNetAnalyzers/StyleCopAnalyzers#1800]
StyleCop is all about consistency. Having empty lines at the end of some files and not having it in others break the consistency. That's why I am suggesting this rule. 

**Category:** Layout rules
**Title:** Code files should/shouldn't end with an empty line

Code without an empty line at the end:
```cs
namespace Namespace
{
    public class Class
    {
    }
}
```

Code with an empty line at the end:
```cs
namespace Namespace
{
    public class Class
    {
    }
}


```

This rule is very similar to https://github.com/DotNetAnalyzers/StyleCopAnalyzers/blob/master/documentation/SA1518.md but 1518 is triggered only when 2 or more end lines are present at the end of the file:

```cs
namespace Namespace
{
    public class Class
    {
    }
}



```

Discussion about ending with an empty line: http://stackoverflow.com/questions/729692/why-should-text-files-end-with-a-newline

GCC also reports this: http://gcc.gnu.org/ml/gcc/2003-11/msg01568.html

**Comment by vweijsters** [comment]
:+1: I agree on the consistency part. We should have the option to choose between either style in the `stylecop.json` settings.

**Comment by sharwell** [comment]
I would prefer to see this implemented as a configuration option for SA1518.

**Comment by nbarbettini** [comment]
:+1: as a configuration option for SA1518.

**Comment by GregReddick** [comment]
:+1: I have this as a rule in my style guide (now published on Amazon). I think in previous versions of Visual Studio (I don't have one handy to check right now, but I seem to recall) that the extra newline showed as an extra line in the Code Metric Results. I just check VS2015, though and it doesn't count one way or the other.

**Comment by 0x084E** [comment]
:+1: for SA1518 option

**Comment by margusmartsepp** [comment]
Just for info, in order to turn on "code must end with a single newline":

1. turn on SA1518 
2. edit stylecop.json and add:

```
        "layoutRules": {
	        "newlineAtEndOfFile": "require"
        } 
```

3. compile project

A use case: you use TeamCity, DotTrace and Sonar together. 👍
